### ISSUE TITLE
Redundant kernel args

### ISSUE THREAD

**Issue by TommyTran732** [Kicksecure/security-misc#199]
According to the kernel's documentation:

auto,nosmt:
Mitigate all CPU vulnerabilities, disabling SMT if needed.  This is for users who always want to be fully mitigated, even if it means losing SMT. Equivalent to: 
- l1tf=flush,nosmt [X86] 
- mds=full,nosmt [X86] 
- tsx_async_abort=full,nosmt [X86] 
- mmio_stale_data=full,nosmt [X86] 
- retbleed=auto,nosmt [X86]

Why are these other args being explicitly set in /etc/grub.d if mitigations=auto,nosmt is already being set?

There is a limit on number of characters that can be in the kernel args as well, and the kernel args we set will just get longer and longer over time. I don't think it is a good idea to waste the precious character on these redundant args. Either we use mitigations=auto,smt (which should be the default anyways), or explicitly spell out which args to set so the user can easily customize them. There is really no reason to have both.

**Comment by adrelanos** [comment]
Saving space in kernel command line, reducing command line options, avoiding redundancy would be good.

We might comment these options out and document that inside the settings file instead out deleting these to avoid having this topic come up again.

Waiting a few days before I look into this and welcoming other comments or review.

**Comment by monsieuremre** [comment]
> auto,nosmt
                                Mitigate all CPU vulnerabilities, disabling SMT
                                if needed.  This is for users who always want to
                                be fully mitigated, even if it means losing SMT.
                                Equivalent to:
l1tf=flush,nosmt [X86]
                                               mds=full,nosmt [X86]
                                               tsx_async_abort=full,nosmt [X86]
                                               mmio_stale_data=full,nosmt [X86]
                                               retbleed=auto,nosmt [X86]
>

Quoting from the documentation. So the man is right. Might as well be less redundant.

**Comment by TommyTran732** [comment]
Should I add a note explicitly saying that these are redundant?

**Comment by adrelanos** [comment]
Yes. That would be nice so anyone else looking into this knows it and no need to rehash this discussion. Always good that way.

**Comment by TommyTran732** [comment]
Well the current default is l1d_flush=on and I am not removing it anyways. I am not advocating for removing everything and only keeping `mitigations=auto,nosmt`, I am advocating for removing the ones that are known to be redundant.

**Comment by TommyTran732** [comment]
> If the default mitigations=auto,nosmt settings were to change and become more lenient in the future, our hardcoded settings would remain fixed and be applied regardless.

Well on this point, we are supposed to check what the kernel does as time goes on anyways. 

> Unless we are currently in need for more characters for kernel arguments, I do not see opening up this potential exposure is necessary?

Another point is that it makes management easier. Having a gigantic list of kargs does make it more painful to deal with when  you need to do debugging and stuff.

I did have a quick talk with Madaidan yesterday and he just said he personally isn't sure if the whole hard coding kargs matters much.

**Comment by adrelanos** [comment]
Useful would be a feature request against the kernel:
`mitigations=strict` should always enable the most secure kernel mitigation settings.

If such a feature were to exist, then no such work downstream in security-misc would be needed.

----

Lower maintenance effort is preferable.

> Should I add a note explicitly saying that these are redundant?

This for sure is actionable. I wouldn't mind more comments saying "this options may be superfluous because...". Or 1 longer comment explaining this and then just a tag "## potentially-superfluous".

Then as a follow-up step the potentially superfluous kernel parameters could even be moved to a separate file. Thereby making it easier for those concerned or impacted by the length of the kernel command line to easily delete these.

**Comment by adrelanos** [comment]
> * https://github.com/Kicksecure/security-misc/pull/320

Is there a way to actually test the effect of using `mitigations=auto,nosmt` versus not using `mitigations=auto,nosmt`?

Idea:

'''1.''' Copy the status of vulnerabilities folder while using `mitigations=auto,nosmt`.

    cp -r /sys/devices/system/cpu/vulnerabilities ~/with

'''2.'''  Disable use of `mitigations=auto,nosmt`.

(Note to other readers: Don't forget `sudo update-grub`, `reboot` and confirm if the kernel parameter change has been actually applied by using `cat /proc/cmdline`.)

'''3.''' Make another copy.

    cp -r /sys/devices/system/cpu/vulnerabilities ~/without

'''4.''' Compare the vulnerabilities folder.

    meld ~/with ~/without

Using that method, can you find any effect?

**Comment by adrelanos** [comment]
As long as the last kernel parameter wins, keeping `mitigations=auto,nosmt` at the top might be best?

Would be useful to point that out explicitly so we don't move `mitigations=auto,nosmt` further down in the future.

related:
https://unix.stackexchange.com/questions/544224/evaluation-order-of-duplicated-kernel-parameters

**Comment by raja-grewal** [comment]
> Well, almost, because all of the other parameters will still be enabled, and theoretically one of those will disable SMT, right?

I am assuming in this experiment that they are all off and we are just using `mitigations=auto` versus `mitigations=auto,nosmt` and nothing else.

> I do not agree that we should disable mitigations=auto,nosmt though. 

Note that `mitigations=auto` is the kernel default and so is redundant, this then combined with our usage of `nosmt=force` is then identical to setting `mitigations=auto,nosmt`.

There is no future mitigation at the default level we are going to miss with this approach. I am not really proposing to commenting it out it for purposes of saving characters, rather, on the simple principle that it serves no purpose.

**Comment by ArrayBolt3** [comment]
> Personally, I disagree with this proposal but at the same time the understand argument that if root is achieved there is little point.

For systems that don't use Secure Boot or kernel lockdown, sure. For systems that *do*... this could be a problem since lockdown mode doesn't allow even root the ability to read arbitrary kernel memory AFAIK (`/proc/kcore` doesn't exist), but if root can just do the equivalent of `mitigations=off`, then it can leak whatever it wants via a side-channel attack. New CPUs are produced today with vulnerabilities that are only avoided by kernel mitigations, so this is a real threat in those scenarios.

I wonder if anyone else has brought that up already, or if someone should bring that up...

**Comment by adrelanos** [comment]
> Note that `mitigations=auto` is the kernel default and so is redundant, this then combined with our usage of `nosmt=force` is then identical to setting `mitigations=auto,nosmt`.

Are we sure that is is really the same? That's hard or impossible to know for sure without looking at the source code.

This argument should therefore still be applicable:

> I do not agree that we should disable mitigations=auto,nosmt though. It is probably dangerous to move down the list, but without it, if a new vulnerability comes out and the kernel adds a mitigation for it, we might not get it if we miss it and don't use mitigations=auto,nosmt. If we do use it, we'll probably get that mitigation, even if we could (and in the future should) set a stronger mitigation on it. I don't think it's too likely we and everyone else will miss a new CPU vulnerability showing up, but if we can harden the system against our own possible ignorance, we should.

Another argument is that `mitigations=auto,nosmt` is probably more widely in use and tested as it's the maximum security setting that one would set based on reading the kernel manual.

In doubt, wasting a bit kernel command line space using `mitigations=auto,nosmt` seems safer than not using it. This solution will be good enough until the kernel completely revises how how to enable maximum mitigation more easily?

**Comment by raja-grewal** [comment]
Apologies for the delay.

You make a good points that in terms of auditing that people would more likely expect this parameter to be present which is also why I wanted to highlight the it's used was flawed in the first place. Regardless, you are correct that until the kernel completely revises how to enable maximum mitigations more clearly we are left in the current scenario.

We can either comment it out as it does not functionally do anything other than in my personal opinion basically spread quite corrosive misinformation. Or as you suggest, simply leave it in place as is and wait for a better solution from upstream.

Therefore, after thinking about this I understand your perspective and respect your decision to keep it active and in place. If your opinion remains the same, should I make the appropriate changes to the PR? Any thoughts @ArrayBolt3?

As an aside, note that even the [attack vector controls](https://docs.kernel.org/admin-guide/hw-vuln/attack_vector_controls.html) introduced in kernel 6.17 appear to inherit the same application of default mitigations. A good future solution would be as you suggested in https://github.com/Kicksecure/security-misc/issues/199#issuecomment-2135165951 to have a sort of `mitigations=strict,nosmt` option.

**Comment by ArrayBolt3** [comment]
Sorry to accidentally leave this for so long, been caught up in a lot of other things.

@raja-grewal Yes, I think keeping the parameter in place would be a good idea. It may be that other settings and kernel configuration that exists make this parameter redundant and even misinformative, but it would only take one bad refactor in the kernel to change how things currently work and possibly leave us in a vulnerable state without the parameter. OpenSSH has had one such bad refactor, a critical PGP library used by Thunderbird had another just recently, I don't trust the kernel to not have the same mistake made with it.

**Comment by ArrayBolt3** [comment]
That does look good to me. I'll do a review of that PR now and most likely merge it in. Thank you!
